x-pack/solutions/observability/plugins/apm_data_access/server/utils/access_known_fields.ts

```diff
@@ -80,30 +99,43 @@ type ApmDocument<
  *
  * console.log(unflattened.service.name); // => outputs "node-svc" like above
  */
-export function accessKnownApmEventFields<
-  T extends Partial<FlattenedApmEvent>,
-  R extends keyof FlattenedApmEvent = never
->(fields: T, required?: R[]): ApmDocument<T, R>;
+export function accessKnownApmEventFields<T extends Partial<FlattenedApmEvent>>(
+  fields: T
+): ApmDocument<T, never>;
 
-export function accessKnownApmEventFields(fields: Record<string, any>, required?: string[]) {
-  if (required) {
-    ensureRequiredApmFields(fields, required);
-  }
+export function accessKnownApmEventFields(fields: Record<string, any>) {
+  const proxy = new Proxy(fields, accessHandler);
 
-  return new Proxy(fields, accessHandler);
+  // Methods are lazily initialised so you only pay the "cost" of what you use.
+  PROXIED_METHODS.set(proxy, {});
+
+  return proxy;
 }
 
 const accessHandler = {
-  get(fields: Record<string, any>, key: string) {
-    if (key === 'unflatten') {
-      return () => unflattenKnownApmEventFields(fields);
-    }
+  get(fields: Record<string, any>, key: string, proxy: any) {
+    switch (key) {
+      case 'unflatten':
+        // Lazily initialise method on first access
+        return (PROXIED_METHODS.get(proxy)!.unflatten ??= () =>
```

Comment:
To keep the initial creation of the proxy lean, the proxied document methods like `unflatten` and `requireFields` are lazily evaluated, so you only pay the cost of initialisation of the methods you use. If the API surface grows, it won't bloat all downstream uses of `accessKnownApmEventFields`.

Reply:
If somehow the `proxy` object is cleaned up just before when this accessHandler is invoked, but that would indicate a bug with the GC cleaning up an object that is still in use.